internal/daemon/api_logs.go

```diff
@@ -0,0 +1,275 @@
+
+type serviceManager interface {
+	Services(names []string) ([]*servstate.ServiceInfo, error)
+	ServiceLogs(services []string, last int) (map[string]servicelog.Iterator, error)
+}
+
+func v1GetLogs(cmd *Command, _ *http.Request, _ *userState) Response {
+	return logsResponse{
+		svcMgr: cmd.d.overlord.ServiceManager(),
+	}
+}
+
+// logsResponse is a Response implementation to serve the logs in a custom
+// JSON Lines format.
+type logsResponse struct {
+	svcMgr serviceManager
+}
+
+func (r logsResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	query := req.URL.Query()
+
+	services := query["services"]
+
+	followStr := query.Get("follow")
+	if followStr != "" && followStr != "true" && followStr != "false" {
+		response := statusBadRequest(`follow parameter must be "true" or "false"`)
+		response.ServeHTTP(w, req)
+		return
+	}
+	follow := followStr == "true"
+
+	numLogs := defaultNumLogs
+	nStr := query.Get("n")
+	if nStr != "" {
+		n, err := strconv.Atoi(nStr)
+		if err != nil {
+			response := statusBadRequest("n must be a valid integer")
+			response.ServeHTTP(w, req)
+			return
+		}
+		numLogs = n
+	}
+
+	// If "services" parameter not specified, fetch logs for all services.
+	if len(services) == 0 {
+		infos, err := r.svcMgr.Services(nil)
+		if err != nil {
+			response := statusInternalError("cannot fetch services: %v", err)
+			response.ServeHTTP(w, req)
+			return
+		}
+		services = make([]string, len(infos))
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service name (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+
+	// First write buffered logs
+	if numLogs != 0 {
+		if numLogs < 0 {
+			// Client asked for all logs; write everything from the iterators
+			// (which are TailIterators).
+			err = writeLogs(itsByName, func(entry servicelog.Entry) error {
+				return encoder.Encode(jsonLog(entry))
+			})
+		} else {
+			// Client asked for n logs, collect from iterators (which are
+			// HeadIterator(n)) to a FIFO queue of size n and drop oldest. This
+			// avoids reading numServices*numLogs into memory and sorting.
+			fifo := make(chan servicelog.Entry, numLogs)
+			err = writeLogs(itsByName, func(entry servicelog.Entry) error {
+				// If FIFO channel is full, discard oldest log entry before
+				// writing new one so it doesn't block.
+				if len(fifo) == cap(fifo) {
+					<-fifo
+				}
+				fifo <- entry
```

Comment:
Note: I'm actually using this channel as a deque / ringbuffer, which has some overhead due to the channel synchronization code. It's about 6x as slow as a hand-rolled deque implementation (see results at https://play.golang.org/p/l09jd5ovdZG), but we're still only talking about ~60ns per append operation (compared to ~10ns) so I prefer the simplicity of just using a channel here.